packages/client/src/compatibility.ts

```diff
@@ -12,126 +12,87 @@ import {
-    private _ctx: RuntimeContext,
-    private _metadataTypes: MetadataTypes,
-    private _descriptors: DescriptorValues,
-  ) {}
-
-  /**
-   * @access package  - Internal implementation detail. Do not use.
-   */
-  static _create(
-    ctx: RuntimeContext,
-    metadataTypes: MetadataTypes,
-    descriptors: DescriptorValues,
-  ) {
-    return new Runtime(ctx, metadataTypes, descriptors)
-  }
-
-  /**
-   * @access package  - Internal implementation detail. Do not use.
-   */
-  _getCtx() {
-    return this._ctx
-  }
-
-  /**
-   * @access package  - Internal implementation detail. Do not use.
-   */
-  _getPalletEntryPoint(
+interface CompatibilityTokenApi {
+  runtime: () => RuntimeContext
+  typedefNodes: TypedefNode[]
+  getPalletEntryPoint: (
     opType: OpType,
     pallet: string,
     name: string,
-  ): EntryPoint {
-    return this._metadataTypes[0][this._descriptors[opType][pallet][name]]
-  }
-
-  /**
-   * @access package  - Internal implementation detail. Do not use.
-   */
-  _getApiEntryPoint(name: string, method: string) {
-    return this._metadataTypes[0][this._descriptors.apis[name][method]]
-  }
-
-  /**
-   * @access package  - Internal implementation detail. Do not use.
-   */
-  _getTypedefNodes() {
-    return this._metadataTypes[1]
-  }
+  ) => EntryPoint
+  getApiEntryPoint: (name: string, method: string) => EntryPoint
 }
+const compatibilityTokenApi = new WeakMap<
+  CompatibilityToken,
+  CompatibilityTokenApi
+>()
+export const getCompatibilityApi = (token: CompatibilityToken) =>
+  compatibilityTokenApi.get(token)!
 
-export type RuntimeApi = Observable<Runtime> & {
-  /**
-   * @returns Promise that resolves in the `Runtime` as soon as it's
-   *          loaded.
-   */
-  latest: () => Promise<Runtime>
+export const enum OpType {
+  Storage = "storage",
+  Tx = "tx",
+  Event = "events",
+  Error = "errors",
+  Const = "constants",
 }
 
 const EntryPointsCodec = Vector(EntryPointCodec)
 const TypedefsCodec = Vector(TypedefCodec)
 const TypesCodec = Tuple(EntryPointsCodec, TypedefsCodec)
-type MetadataTypes = ReturnType<(typeof TypesCodec)["dec"]>
-const decodedMetadataTypes = new WeakMap<Uint8Array, MetadataTypes>()
 
-export const getRuntimeApi = (
-  metadataTypes: Promise<Uint8Array>,
-  descriptors: Promise<DescriptorValues>,
+const tokens = new WeakMap<ChainDefinition, Promise<CompatibilityToken<any>>>()
+export const createCompatibilityToken = <D extends ChainDefinition>(
+  chainDefinition: D,
   chainHead: ReturnType<ReturnType<typeof getObservableClient>["chainHead$"]>,
-): RuntimeApi => {
-  const decodedMetadata = metadataTypes.then((byteArr) => {
-    if (decodedMetadataTypes.has(byteArr)) {
-      return decodedMetadataTypes.get(byteArr)!
-    }
-    const result = TypesCodec.dec(byteArr)
-    decodedMetadataTypes.set(byteArr, result)
-    return result
+): Promise<CompatibilityToken<D>> => {
+  if (tokens.has(chainDefinition)) {
+    return tokens.get(chainDefinition)!
+  }
```

Comment:
This is actually wrong.... I can't use the `chainDefinition` as a key for a map that's shared for all clients.

This wouldn't work properly:
```ts
const chainAClient = createClient(WebSocketProvider(urlA));
const chainBClient = createClient(WebSocketProvider(urlB));

// Creates the entry in this `tokens` map with `descriptors => something that listens on `chainA`'s chainHead
const typedApiA = chainAClient.getTypedApi(descriptors);
// Wrong: reuses the same token from `typedApiA`, which is listening to `chainA`'s chainHead.
const typedApiB = chainBClient.getTypedApi(descriptors);
```